src/devices/vesternet.ts

```diff
@@ -13,9 +14,15 @@ const definitions: Definition[] = [
         vendor: 'Vesternet',
         description: 'Zigbee dimmer',
         fromZigbee: extend.light_onoff_brightness().fromZigbee
-            .concat([fz.electrical_measurement, fz.metering, fz.ignore_genOta]),
-        toZigbee: extend.light_onoff_brightness().toZigbee.concat([tz.power_on_behavior]),
-        exposes: [e.light_brightness(), e.power(), e.voltage(), e.current(), e.energy(), e.power_on_behavior(['off', 'on', 'previous'])],
+            .concat([fz.electrical_measurement, fz.metering, fz.ignore_genOta, fz.level_config]),
+        toZigbee: extend.light_onoff_brightness().toZigbee.concat([tz.power_on_behavior, tz.level_config]),
+        exposes: [e.light_brightness(), e.power(), e.voltage(), e.current(), e.energy(), e.power_on_behavior(['off', 'on', 'previous']),
```

Comment:
Instead of `e.composite('level_config')`, can you try `e.light_brightness().withLevelConfig()`?

Reply:
hi thx you for your reply, i test with external but withLevelConfig no contains the on_level

```
  withLevelConfig() {
        assert(!this.endpoint, 'Cannot add feature after adding endpoint');
        const levelConfig = new Composite('level_config', 'level_config', access.ALL)
            .withFeature(new Numeric('on_off_transition_time', access.ALL)
                .withLabel('ON/OFF transition time')
                .withDescription('Represents the time taken to move to or from the target level when On of Off commands are received by an On/Off cluster'),
            )
            .withFeature(new Numeric('on_transition_time', access.ALL)
                .withLabel('ON transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the minimum level to the maximum level when an On command is received'),
            )
            .withFeature(new Numeric('off_transition_time', access.ALL)
                .withLabel('OFF transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the maximum level to the minimum level when an Off command is received'),
            )
            .withFeature(new Numeric('current_level_startup', access.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('minimum', 0, 'Use minimum permitted value')
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Defines the desired startup level for a device when it is supplied with power'),
            )
            .withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);

        return this;
    }
```

![image](https://github.com/Koenkk/zigbee-herdsman-converters/assets/8700106/ce291ba4-d015-4bd0-b0ad-da1bc23aa1da)

can i override/add the withLevelConfig function or extend the class Light in external converter ?

Reply:
I could create a function in the Light class with the name withOnLevel, but I wouldn't be able to test it and I don't like that ^^

Reply:
ok i validate with that 

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;


class LightWithOnLevel extends exposes.Light {

	 withOnLevel() {
        const levelConfig = e.composite('level_config', 'level_config', ea.ALL)
            .withFeature(e.numeric('on_level', ea.ALL)
				.withValueMin(1).withValueMax(254)
				.withPreset('previous', 255, 'Use previous value')
				.withDescription('Specifies the level that shall be applied, when an on/toggle command causes the light to turn on.'))
            .withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);

        return this;
    }

}


const definition = {
        fingerprint: [{modelID: 'HK-SL-DIM-A', softwareBuildID: '2.5.3_r52'}, {modelID: 'HK-SL-DIM-A', softwareBuildID: '2.9.2_r54'}],
        model: 'VES-ZB-DIM-004',
        vendor: 'Vesternet',
        description: 'Zigbee dimmer',
        fromZigbee: extend.light_onoff_brightness().fromZigbee
            .concat([fz.electrical_measurement, fz.metering, fz.ignore_genOta, fz.level_config]),
        toZigbee: extend.light_onoff_brightness().toZigbee.concat([tz.power_on_behavior, tz.level_config]),
        exposes: [new LightWithOnLevel().withBrightness().withOnLevel(),
					e.power(), e.voltage(), e.current(), e.energy(), e.power_on_behavior(['off', 'on', 'previous'])
        ],
        whiteLabel: [{vendor: 'Sunricher', model: 'SR-ZG9040A'}],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'haElectricalMeasurement', 'seMetering']);
            await reporting.onOff(endpoint);
            await reporting.brightness(endpoint);
            await reporting.readEletricalMeasurementMultiplierDivisors(endpoint);
            await reporting.activePower(endpoint);
            await reporting.rmsCurrent(endpoint, {min: 10, change: 10});
            await reporting.rmsVoltage(endpoint, {min: 10});
            await reporting.readMeteringMultiplierDivisor(endpoint);
            await reporting.currentSummDelivered(endpoint);
        },
    }
module.exports = definition;
```


What do you prefer, modifying the light class directly or trying to integrate it into the device's conf?

Reply:
hi @Koenkk , finally I create the extension class directly on the vesternet file because it is device specific, for example execute_if_off, is present but enables or disables the level control and not directly the brightness

to solve the problem with the pipline, I deleted fz.level_config and tz.level_control (for fromZigbee and ToZigbee respectively), it seems to work but I don't understand why...

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;


class LightWithOnLevel extends exposes.Light {

	 withOnLevel() {
        const levelConfig = e.composite('level_config', 'level_config', ea.ALL)
                .withFeature(e.numeric('on_off_transition_time', ea.ALL)
                    .withDescription('Specifies the amount of time, in units of 0.1 seconds, which will be used during a transition to ' +
                    'either the on or off state, when an on/off/toggle command of the on/off cluster is used to turn the light on or off'))
                .withFeature(e.binary('execute_if_off', ea.ALL, true, false)
                    .withLabel('Enable level control').withDescription('this parameter activates or deactivates the "on_level" and "current_level_startup" features'))
                .withFeature(e.numeric('on_level', ea.ALL)
                    .withValueMin(1).withValueMax(254)
                    .withPreset('previous', 255, 'Use previous value')
                    .withDescription('Specifies the level that shall be applied, when an on/toggle command causes the light to turn on.'))
				.withFeature(e.numeric('current_level_startup', ea.ALL)
                    .withValueMin(1).withValueMax(254)
                    .withPreset('previous', 255, 'Use previous value')
                    .withDescription('Specifies the initial level to be applied after the device is supplied with power'))
            .withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);

        return this;
    }
}


const definition = {
        fingerprint: [{modelID: 'HK-SL-DIM-A', softwareBuildID: '2.5.3_r52'}, {modelID: 'HK-SL-DIM-A', softwareBuildID: '2.9.2_r54'}],
        model: 'VES-ZB-DIM-004',
        vendor: 'Vesternet',
        description: 'Zigbee dimmer',
        fromZigbee: extend.light_onoff_brightness().fromZigbee
            .concat([fz.electrical_measurement, fz.metering, fz.ignore_genOta]),
        toZigbee: extend.light_onoff_brightness().toZigbee.concat([tz.power_on_behavior]),
        exposes: [new LightWithOnLevel().withBrightness().withOnLevel(),
					e.power(), e.voltage(), e.current(), e.energy(), e.power_on_behavior(['off', 'on', 'previous'])
        ],
        whiteLabel: [{vendor: 'Sunricher', model: 'SR-ZG9040A'}],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'haElectricalMeasurement', 'seMetering']);
            await reporting.onOff(endpoint);
            await reporting.brightness(endpoint);
            await reporting.readEletricalMeasurementMultiplierDivisors(endpoint);
            await reporting.activePower(endpoint);
            await reporting.rmsCurrent(endpoint, {min: 10, change: 10});
            await reporting.rmsVoltage(endpoint, {min: 10});
            await reporting.readMeteringMultiplierDivisor(endpoint);
            await reporting.currentSummDelivered(endpoint);
        },
    }
module.exports = definition;
```

I was able to validate with an external file, however I don't know if I should put extend for the LightWithOnLevel class,  i put the comment direct on the code if you can check :)

Reply:
result :

![image](https://github.com/Koenkk/zigbee-herdsman-converters/assets/8700106/ea2de27a-fab9-42da-9a45-6f493042e08b)

Reply:
Please add the `on_level` to `withLevelConfig` instead

Reply:
hi @Koenkk, my problem with  'withLevelConfig' contains on_transition_time and off_transition_time, but not existing on my device :

```
"level_config": {
        "current_level_startup": "previous",
        "execute_if_off": true,
        "on_level": 176,
        "on_off_transition_time": 0
    },
```

that why i create extend class, but do you want me to directly create a new function in the light class?

Reply:
proposal : 

```
    withLevelConfig(customIncludes: string[] = []) {
        assert(!this.endpoint, 'Cannot add feature after adding endpoint');
        const levelConfig = new Composite('level_config', 'level_config', access.ALL);
        if (customIncludes.length == 0 || customIncludes.includes('on_off_transition_time')) {
            levelConfig.withFeature(new Numeric('on_off_transition_time', access.ALL)
                .withLabel('ON/OFF transition time')
                .withDescription('Represents the time taken to move to or from the target level when On of Off commands are received by an On/Off cluster'),
            )
        }
        if (customIncludes.length == 0 || customIncludes.includes('on_transition_time')) {
            levelConfig.withFeature(new Numeric('on_transition_time', access.ALL)
                .withLabel('ON transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the minimum level to the maximum level when an On command is received'),
            );
        }
        if (customIncludes.length == 0 || customIncludes.includes('off_transition_time')) {
            levelConfig.withFeature(new Numeric('off_transition_time', access.ALL)
                .withLabel('OFF transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the maximum level to the minimum level when an Off command is received'),
            );
        }
        if (customIncludes.length == 0 || customIncludes.includes('current_level_startup')) {
            levelConfig.withFeature(new Numeric('current_level_startup', access.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('minimum', 0, 'Use minimum permitted value')
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Defines the desired startup level for a device when it is supplied with power'),
            );
        }
        if (customIncludes.includes('execute_if_off')) {
            levelConfig.withFeature(new Binary('execute_if_off', access.ALL, true, false)
                .withDescription('this parameter activates or deactivates the "on_level" and "current_level_startup" features and force level to 1'));
        }
        if (customIncludes.includes('on_level')) {
            levelConfig.withFeature(new Numeric('on_level', access.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Specifies the level that shall be applied, when an on/toggle command causes the light to turn on.'))
        }
        levelConfig.withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);

        return this;
    }

```

Reply:
I propose to create an `disableFeatures` instead, then the logic can be:

`if (!disableFeatures.includes('current_level_startup')) {`

Reply:
> I propose to create an `disableFeatures` instead, then the logic can be:
> 
> `if (!disableFeatures.includes('current_level_startup')) {`

I can but in this case I will have to modify all devices that use withLevelConfig, this can make a lot of modifications, that's why I did that.
what do you choose?
- an exclusion list with a modification of all devices that use withLevelConfig
- an inclusion list which if it is empty, leaves the initial behavior and if it is supplied allows customization, without having to modify the other devices

Reply:
Why does an exclusion list require modifications to other devices? By default nothing is excluded which we want for all devices except this one.

Reply:
> Why does an exclusion list require modifications to other devices? By default nothing is excluded which we want for all devices except this one.

presently ti add 2 elements on withLevelConfig: 
- execute_if_off
- on_level

with exclusion list all devices that already implement withLevelConfig will need to be modified to exclude both of its elements.

Currently with my inclusion list, I set the original values ​​by default if no parameter is passed. and if we pass parameters in this case I only push the chosen elements

Reply:
I could do the same thing with the exclusion list but it complicates the code and can cause comprehension errors

```
    withLevelConfig(disableFeatures: string[] = []) {
        assert(!this.endpoint, 'Cannot add feature after adding endpoint');
        const levelConfig = new Composite('level_config', 'level_config', access.ALL);
        if (disableFeatures.length == 0 || !disableFeatures.includes('on_off_transition_time')) {
            levelConfig.withFeature(new Numeric('on_off_transition_time', access.ALL)
                .withLabel('ON/OFF transition time')
                .withDescription('Represents the time taken to move to or from the target level when On of Off commands are received by an On/Off cluster'),
            )
        }
        if (disableFeatures.length == 0 || !disableFeatures.includes('on_transition_time')) {
            levelConfig.withFeature(new Numeric('on_transition_time', access.ALL)
                .withLabel('ON transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the minimum level to the maximum level when an On command is received'),
            );
        }
        if (disableFeatures.length == 0 || !disableFeatures.includes('off_transition_time')) {
            levelConfig.withFeature(new Numeric('off_transition_time', access.ALL)
                .withLabel('OFF transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the maximum level to the minimum level when an Off command is received'),
            );
        }
        if (disableFeatures.length == 0 || !disableFeatures.includes('current_level_startup')) {
            levelConfig.withFeature(new Numeric('current_level_startup', access.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('minimum', 0, 'Use minimum permitted value')
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Defines the desired startup level for a device when it is supplied with power'),
            );
        }
        if (disableFeatures.length > 0 && !disableFeatures.includes('execute_if_off')) {
            levelConfig.withFeature(new Binary('execute_if_off', access.ALL, true, false)
                .withDescription('this parameter activates or deactivates the "on_level" and "current_level_startup" features and force level to 1'));
        }
        if (disableFeatures.length > 0 && !disableFeatures.includes('on_level')) {
            levelConfig.withFeature(new Numeric('on_level', access.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Specifies the level that shall be applied, when an on/toggle command causes the light to turn on.'));
        }
        levelConfig.withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);

        return this;
    }
```

Reply:
exemple with another devices : 

```
  {
        zigbeeModel: ['PUCK/DIMMER/1'],
        model: 'CCT5010-0001',
        vendor: 'Schneider Electric',
        description: 'Micro module dimmer',
        fromZigbee: [...extend.light_onoff_brightness().fromZigbee, fz.wiser_lighting_ballast_configuration],
        toZigbee: [...extend.light_onoff_brightness().toZigbee, tz.ballast_config, tz.wiser_dimmer_mode],
        exposes: [e.light_brightness().withLevelConfig(),
            e.numeric('ballast_minimum_level', ea.ALL).withValueMin(1).withValueMax(254)
                .withDescription('Specifies the minimum light output of the ballast'),
            e.numeric('ballast_maximum_level', ea.ALL).withValueMin(1).withValueMax(254)
                .withDescription('Specifies the maximum light output of the ballast'),
            e.enum('dimmer_mode', ea.ALL, ['auto', 'rc', 'rl', 'rl_led'])
                .withDescription('Sets dimming mode to autodetect or fixed RC/RL/RL_LED mode (max load is reduced in RL_LED)')],
        whiteLabel: [{vendor: 'Elko', model: 'EKO07090'}],
        configure: async (device, coordinatorEndpoint, logger) => {
            await extend.light_onoff_brightness().configure(device, coordinatorEndpoint, logger);
            const endpoint = device.getEndpoint(3);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'lightingBallastCfg']);
            await reporting.onOff(endpoint);
            await reporting.brightness(endpoint);
        },
    },
```
this device already implements withLevelConfig but I don't know if it has the execute_if_off and on_level attributes, so I should add an exclusion list for 'execute_if_off' and 'on_level'

Reply:
I just looked that's 8 occurrences of using 'withLevelConfig'
I can do it if you want

Reply:
I expect that all devices that are using the `tz.level_config` support `execute_if_off` and `on_level`, so it's fine if those get added to the other devices.

Reply:
sorry for lot of commit, i can commit direct on intellij i have an issue with lint, and apply the change direct on github...
i cannot test the change but normally it's OK

Reply:
I tested with the external file, it seems to work
```

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;


class LightWithOnLevel extends exposes.Light {

	 withOnLevel(disableFeatures = []) {
         let levelConfig = e.composite('level_config', 'level_config', ea.ALL);
        if (!disableFeatures.includes('on_off_transition_time')) {
            levelConfig = levelConfig.withFeature(e.numeric('on_off_transition_time', ea.ALL)
                .withLabel('ON/OFF transition time')
                .withDescription('Represents the time taken to move to or from the target level when On of Off commands are received by an On/Off cluster'));
        }
        if (!disableFeatures.includes('on_transition_time')) {
            levelConfig = levelConfig.withFeature(e.numeric('on_transition_time', ea.ALL)
                .withLabel('ON transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the minimum level to the maximum level when an On command is received'));
        }
        if (!disableFeatures.includes('off_transition_time')) {
            levelConfig = levelConfig.withFeature(e.numeric('off_transition_time', ea.ALL)
                .withLabel('OFF transition time')
                .withPreset('disabled', 65535, 'Use on_off_transition_time value')
                .withDescription('Represents the time taken to move the current level from the maximum level to the minimum level when an Off command is received'));
        }
        if (!disableFeatures.includes('execute_if_off')) {
            levelConfig = levelConfig.withFeature(e.binary('execute_if_off', ea.ALL, true, false)
                .withDescription('this setting can affect the "on_level", "current_level_startup" or "brightness" setting'));
        }
        if (!disableFeatures.includes('on_level')) {
            levelConfig = levelConfig.withFeature(e.numeric('on_level', ea.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Specifies the level that shall be applied, when an on/toggle command causes the light to turn on.'));
        }
        if (!disableFeatures.includes('current_level_startup')) {
            levelConfig = levelConfig.withFeature(e.numeric('current_level_startup', ea.ALL)
                .withValueMin(1).withValueMax(254)
                .withPreset('minimum', 0, 'Use minimum permitted value')
                .withPreset('previous', 255, 'Use previous value')
                .withDescription('Defines the desired startup level for a device when it is supplied with power'));
        }
        levelConfig = levelConfig.withDescription('Configure genLevelCtrl');
        this.features.push(levelConfig);
        return this;
	 }
}


const definition = {
        fingerprint: [{modelID: 'HK-SL-DIM-A', softwareBuildID: '2.5.3_r52'}, {modelID: 'HK-SL-DIM-A', softwareBuildID: '2.9.2_r54'}],
        model: 'VES-ZB-DIM-004',
        vendor: 'Vesternet',
        description: 'Zigbee dimmer',
        fromZigbee: extend.light_onoff_brightness().fromZigbee
            .concat([fz.electrical_measurement, fz.metering, fz.ignore_genOta]),
        toZigbee: extend.light_onoff_brightness().toZigbee.concat([tz.power_on_behavior]),
        exposes: [new LightWithOnLevel().withBrightness().withOnLevel(['on_transition_time', 'off_transition_time']),
					e.power(), e.voltage(), e.current(), e.energy(), e.power_on_behavior(['off', 'on', 'previous'])
        ],
        whiteLabel: [{vendor: 'Sunricher', model: 'SR-ZG9040A'}],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint = device.getEndpoint(1);
            await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl', 'haElectricalMeasurement', 'seMetering']);
            await reporting.onOff(endpoint);
            await reporting.brightness(endpoint);
            await reporting.readEletricalMeasurementMultiplierDivisors(endpoint);
            await reporting.activePower(endpoint);
            await reporting.rmsCurrent(endpoint, {min: 10, change: 10});
            await reporting.rmsVoltage(endpoint, {min: 10});
            await reporting.readMeteringMultiplierDivisor(endpoint);
            await reporting.currentSummDelivered(endpoint);
        },
    }
module.exports = definition;
```